### ISSUE TITLE
[ENH] clean implementation of `DeepAptamer`

### ISSUE THREAD

**Issue by fkiraly** [gc-os-ai/pyaptamer#82]
After prioritization, we decided that `DeepAptamer` would be a nice next algorithm to "clean up" for easy use.

[code](https://github.com/YangX-BIDD/DeepAptamer) [paper](https://pubmed.ncbi.nlm.nih.gov/39897584/) [pdf](https://www.cell.com/molecular-therapy-family/nucleic-acids/pdf/S2162-2531(24)00323-8.pdf)

**Comment by satvshr** [comment]
Blocker: `DeepAptamer` uses 4 shape properties which they got from a web server called `DNAshape`, there is an R repository by some authors of that paper called [`DNAshapeR`](https://github.com/TsuPeiChiu/DNAshapeR), there is no python repository for the same. My question is: Is there any way to run R code in python?

**Comment by satvshr** [comment]
@KubiczekD @JaBirke In the paper (page 6) under "Pre-trained DeepAptamer could identify key nucleotides
essential for target binding" they mention VAE (which I think suggests where the aptamer can bind with the target protein) as a use-case. I remember Dennis talking about the use of DeepAptamer for this, but I would like you to know VAE is not a part of the architecture but the output from DeepAptamer is something VAE uses (There is no code on how the VAE is constructed either), just to avoid misunderstandings! Kindly correct me if I am wrong.

**Comment by satvshr** [comment]
* The [paper](https://www.cell.com/molecular-therapy-family/nucleic-acids/pdf/S2162-2531(24)00323-8.pdf) mentions using a 2D convolutional filter in page 4 under "The integrated DeepAptamer model demonstrated superior performance compared with the individual CNN or BiLSTM model" but the [image](https://github.com/YangX-BIDD/DeepAptamer/blob/84ee416702630a4e244c78f92e673e0ddbcda28b/model/BCMA/cnn_bilstm/model.png) mentions using a 1D convolutional model instead.
* The image linked above also includes attention layers for some reason and when you try searching for the word attention in the official paper, it only links to 2 citations in references.

**Comment by fkiraly** [comment]
> DeepAptamer uses 4 shape properties which they got from a web server called DNAshape

What are these as objects? Type, size, etc

**Comment by satvshr** [comment]
> I would read this as "variational auto-encoder", this is a type of generative neural network

I was not doubting what it is, I am confused whether it is a part of DeepAptamer, but I am fairly certain it is not.

> > DeepAptamer uses 4 shape properties which they got from a web server called DNAshape
> 
> What are these as objects? Type, size, etc

I was going to mention it later on but here is where it gets interesting: there are 4 properties but the shape matrix is supposed to be 162 flattened dimensions long, which implies that not all shapes properties have vectors of equal length (given 162 does not get divided by 4)

**Comment by fkiraly** [comment]
@satvshr, is there a python drop-in for these features?

If not, we may want to consider implementing them, they do not look too difficult.

**Comment by satvshr** [comment]
> I would suggest you do _not_ look at the R code, because it is restrictively licensed (GPL, this is copyleft). Instead, look only at how it is used from within python and try to derive specifications.

I just set up everything to start understanding the R repo, I will verify if it is used within python but as I said in the daily, it most likely is a part of the input.

**Comment by fkiraly** [comment]
> My question to @fkiraly: Why is the R package not free-to-use but the python package is?

The R package is subject to the GPL-2 license which is copyleft - this makes it unusable for most commercial purposes https://en.wikipedia.org/wiki/Copyleft

The python package is BSD-3, which is a permissive license (with attribution requirement), same that sktime also has.

**Comment by fkiraly** [comment]
> This is the [webserver](https://rohslab.cmb.usc.edu/DNAshape/) the authors of DeepAptamer used to get their shapes (same authors of DNAshapeR).

A webserver? Why would you need a webserver in a workflow that looks like it can be completely automated?

**Comment by JaBirke** [comment]
As far as I understand, DeepDNAshape and methyl-DNAshape can provide similar functionality to DNAshape/DNAshapeR . I would suggest DeepDNAshape, since it is Python-based and offers a lot of shape features. methyl-DNAshape seems designed specifically for methylated DNA and, according to the GitHub repository (https://github.com/satyanarayan-rao/methyl-DNAshapeR),  is a extension of DNAshapeR, so I assume it is not python-based

**Comment by fkiraly** [comment]
> I do not think we should be trying to implement [deepDNAshape](https://github.com/JinsenLi/deepDNAshape) on our own as:

> Do you agree with what I am saying?

> repo there is no direct way to install it using a simple pip command, so is there 

Yes, this was my answer: https://github.com/gc-os-ai/pyaptamer/issues/82#issuecomment-3189757116

Specifically:

* I recommend to copy the code and remove the CLI interface. Ensure proper attribution by file to comply with BSD-3 license.
* git installs are unsafe and hard to manage, let´s not do that.

**Comment by satvshr** [comment]
> I recommend to copy the code

Apologies, this statement is not very clear to me (or you could be being ambiguous :p) as I by "the code" you could mean:
* Replicate what deepDNAshape is doing in terms of functionality (i.e., retrain their neural networks, build them in torch, and so on)
* Copy the code which they use in their CLI interface to call inbuilt functions.

Currently I am trying to install the deepDNAshape library and it requires python 3.11 along with tensorflow >= 2.6.0 and numpy<1.24. It does not seem very convenient, plus given you said:
> git installs are unsafe and hard to manage, let´s not do that.

It will be hard for the user to set it up. I assume there is no other method of making it easier to use?

**Comment by satvshr** [comment]
> ohhhh, it uses `tensorflow`? Then I would recommend to translate to `torch` instead of copying.

As I said, weights are pre-loaded and not in `.h5` so torch wont support it, whats wrong with having tensorflow as an optional dependency?

**Comment by satvshr** [comment]
I will try to have a look if earlier versions of torch can maybe handle those weight files? It does not seem worth the hassle though, to re-implement most of the library just for `DeepAptamer`, it does not seem easy and will be time-consuming.

**Comment by satvshr** [comment]
@JaBirke @KubiczekD @fkiraly  To give context, we can use [`Deep DNAshape`](https://github.com/JinsenLi/deepDNAshape) and not [`DNAshape`](https://github.com/TsuPeiChiu/DNAshapeR), the question "why?" can be inferred by reading the discussion above. A few observations I had regarding the extraction of shapes in `DNAshape` (I will refer below as A) vs `Deep DNAshape`(I will refer as B) in terms of differences and similarities, **please correct me if I am wrong**:
* Similarities: A and B both use the concept of `pentamers`, A has it hardcoded (no option to adjust a sliding window of 5) whereas B allows an [adjustable sliding window](https://github.com/JinsenLi/deepDNAshape?tab=readme-ov-file#predict-any-dna-shape-for-a-single-sequence), or at least thats what I infer from the `layer` parameter. So, to get a sliding window of 5 we need to input the `layer` parameter as 2.
* In the [paper](https://www.cell.com/molecular-therapy-family/nucleic-acids/pdf/S2162-2531(24)00323-8.pdf) page 11 under "DNA shape features" the way they describe what they did makes no sense.
> To ensure consistency between the input DNA sequence set
and the corresponding DNA shape properties, both sides of the matrix were filled with two zeros, the length was set to l + 4, and a sliding window was used to obtain a shape property matrix n x l.

For starters, they mention getting a shape matrix of size 126, which would just not be possible with the current mentioned method, here is why:
If I input a 35 length long aptamer sequence, I get (from both A and B) vectors of len:
* MGW and PROT: 35
* HElT and Roll: 34
which adds up to 138, now here comes the catch (and the difference between A and B:
These are the outputs for a sequence `GTACGTACGTACGTACGTACGTACGTACGTACGTA`.
For A:
MGW:
```
NA,NA,5.93,4.84,4.84,5.93,5.93,4.84,4.84,5.93,5.93,4.84,4.84,5.93,5.93,4.84
4.84,5.93,5.93,4.84,4.84,5.93,5.93,4.84,4.84,5.93,5.93,4.84,4.84,5.93,5.93,4.84
4.84,NA,NA
```
HelT:
```
NA,34.20,34.13,34.34,34.13,34.20,34.13,34.34,34.13,34.20,34.13,34.34,34.13,34.20,34.13,34.34
34.13,34.20,34.13,34.34,34.13,34.20,34.13,34.34,34.13,34.20,34.13,34.34,34.13,34.20,34.13,34.34
34.30,NA
```
PROT:
```
NA,NA,-8.84,-7.92,-7.92,-8.84,-8.84,-7.92,-7.92,-8.84,-8.84,-7.92,-7.92,-8.84,-8.84,-7.92
-7.92,-8.84,-8.84,-7.92,-7.92,-8.84,-8.84,-7.92,-7.92,-8.84,-8.84,-7.92,-7.92,-8.84,-8.84,-7.92
-7.92,NA,NA
```
Roll:
```
NA,4.68,-2.40,5.09,-2.40,4.68,-2.40,5.09,-2.40,4.68,-2.40,5.09,-2.40,4.68,-2.40,5.09
-2.40,4.68,-2.40,5.09,-2.40,4.68,-2.40,5.09,-2.40,4.68,-2.40,5.09,-2.40,4.68,-2.40,5.09
-2.68,NA
```
There are a total of 12 `NA` values in the 4 shape vectors, so 138 - 12 = 126 (size of the shape vector!). There is no 0 padding taking place as mentioned in the "DNA shape features" section.

For B:
MGW:
```
[5.1615953 5.905737  6.042103  5.0043426 5.0043426 6.042103  6.042103
 5.0043426 5.0043426 6.042103  6.042103  5.0043426 5.0043426 6.042103
 6.042103  5.0043426 5.0043426 6.042103  6.042103  5.0043426 5.0043426
 6.042103  6.042103  5.0043426 5.0043426 6.042103  6.042103  5.0043426
 5.0043426 6.042103  6.042103  5.0043426 5.0043426 5.952573  5.8894873]
```
HelT:
```
[33.516754 34.131874 34.33365  34.368088 34.33365  34.07556  34.33365
 34.368088 34.33365  34.07556  34.33365  34.368088 34.33365  34.07556
 34.33365  34.368088 34.33365  34.07556  34.33365  34.368088 34.33365
 34.07556  34.33365  34.368088 34.33365  34.07556  34.33365  34.368088
 34.33365  34.07556  34.33365  34.368088 34.437675 33.758266]
```
PROT:
```
[-0.8210573 -7.975397  -8.752142  -7.661254  -7.661254  -8.752142
 -8.752142  -7.661254  -7.661254  -8.752142  -8.752142  -7.661254
 -7.661254  -8.752142  -8.752142  -7.661254  -7.661254  -8.752142
 -8.752142  -7.661254  -7.661254  -8.752142  -8.752142  -7.661254
 -7.661254  -8.752142  -8.752142  -7.661254  -7.661254  -8.752142
 -8.752142  -7.661254  -7.661254  -9.289152  -5.226278 ]
```
Roll:
```
[-1.9796891  4.9636946 -2.373889   4.9559116 -2.373889   5.0104446
 -2.373889   4.9559116 -2.373889   5.0104446 -2.373889   4.9559116
 -2.373889   5.0104446 -2.373889   4.9559116 -2.373889   5.0104446
 -2.373889   4.9559116 -2.373889   5.0104446 -2.373889   4.9559116
 -2.373889   5.0104446 -2.373889   4.9559116 -2.373889   5.0104446
 -2.373889   4.9559116 -2.6350822  5.7576814]
```

The good news is that if we remove positions in B where there were `NA` values in A, the shape values are pretty similar!

My questions:
1. Given part of what the authors say in "DNA shape features" is False (at least that is where the evidence shown above leads me), should I also perform the normalization they mentioned in that section?
2. Can I just remove the first and last values in B where there is `NA` (observably, the first and last values in `Roll and Helt` and the first 2 and last 2 values in `MGW and PROT` ) in A?
